src/wallet3/db_schema.cpp

```diff
@@ -420,5 +425,41 @@ namespace wallet
   {
     return prepared_get<int64_t>("SELECT output_count FROM metadata WHERE id=0;");
   }
+  void
+  WalletDB::save_keys(const std::string& spend_priv_str, const std::string& spend_pub_str, const std::string& view_priv_str, const std::string& view_pub_str)
+  {
+    std::string query = "SELECT spend_priv, spend_pub, view_priv, view_pub FROM metadata WHERE id=0";
+
+    auto st = prepared_st(query);
+    st->executeStep();
+    auto from_db = db::get<std::string, std::string, std::string, std::string>(st);
+    const auto loaded_spend_priv_str = std::get<0>(from_db);
+    const auto loaded_spend_pub_str = std::get<1>(from_db);
+    const auto loaded_view_priv_str = std::get<2>(from_db);
+    const auto loaded_view_pub_str = std::get<3>(from_db);
+
+    if ((loaded_spend_priv_str != "" && loaded_spend_priv_str != spend_priv_str) ||
+        (loaded_spend_pub_str != "" && loaded_spend_pub_str != spend_pub_str) ||
+        (loaded_view_priv_str != "" && loaded_view_priv_str != view_priv_str) ||
+        (loaded_view_pub_str != "" && loaded_view_pub_str != view_pub_str))
+          throw std::runtime_error("loaded keys do not match database file");
+
+    prepared_exec("UPDATE metadata SET spend_priv = ?, spend_pub = ?, view_priv = ?, view_pub = ? where id = 0;",
+        spend_priv_str,
+        spend_pub_str,
+        view_priv_str,
+        view_pub_str);
+
+  }
+
+  std::tuple<std::string, std::string, std::string, std::string>
+  WalletDB::load_keys()
+  {
+    std::string query = "SELECT spend_priv, spend_pub, view_priv, view_pub FROM metadata WHERE id=0";
+    auto st = prepared_st(query);
+    st->executeStep();
+    auto from_db = db::get<std::string, std::string, std::string, std::string>(st);
+    return std::make_tuple(std::get<0>(from_db), std::get<1>(from_db), std::get<2>(from_db), std::get<3>(from_db));
```

Comment:
There is way too much complexity here that is avoiding the abstractions that Database provides to avoid this sort of complexity.  The whole thing can be written in one line:
```suggestion
    return prepared_get<std::string, std::string, std::string, std::string>(
        "SELECT spend_priv, spend_pub, view_priv, view_pub FROM metadata WHERE id=0");
```